File: waymo_open_dataset/utils/test_utils.py
# Copyright 2019 The Waymo Open Dataset Authors. All Rights Reserved.
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.
# ==============================================================================
"""Test utils."""

from __future__ import absolute_import
from __future__ import division
from __future__ import print_function

import numpy as np
import tensorflow as tf

from waymo_open_dataset.utils import transform_utils

__all__ = ['generate_boxes', 'generate_range_image', 'generate_extrinsic']


def generate_boxes(center_x, batch=1):
  """Generates unit size boxes centered at (center_x, 0.0, 2.0).

  Args:
    center_x: center_x for each box
    batch: batch size

  Returns:
    boxes: [batch, len(center_x) + 5, 7]. Each box has unit size, zero heading.
      They are centered at (center_x, 0.0, 2.0).
    box_classes: [batch, len(center_x) + 5]. The types are random.
    num_boxes: [batch]. Number of boxes for each batch dim.
  """
  center_y = np.zeros_like(center_x)
  center_z = np.ones_like(center_x) * 2.0
  heading = np.zeros_like(center_x)
  dim = np.ones_like(center_x)

  # [len(center_x), 7]
  boxes = np.stack([center_x, center_y, center_z, dim, dim, dim, heading],
                   axis=-1)

  box_padding = np.zeros([5, 7])

  boxes = np.concatenate([boxes, box_padding], axis=0)
  boxes = np.tile(np.expand_dims(boxes, axis=0), [batch, 1, 1])

  box_classes = np.tile(
      np.expand_dims(np.random.randint(1, 5, size=[len(center_x) + 5]), axis=0),
      [batch, 1])
  num_boxes = np.ones([batch]) * len(center_x)

  return (tf.convert_to_tensor(value=boxes, dtype=tf.float32),
          tf.convert_to_tensor(value=box_classes, dtype=tf.uint8),
          tf.convert_to_tensor(value=num_boxes, dtype=tf.int32))


def generate_range_image(indices, values, shape, batch=1):
  """Generate range images by scattering values to indices.

  Args:
    indices: [N, 2] indices.
    values: [N, ...] values.
    shape: [3].
    batch: batch indices, single integer.

  Returns:
    range_image: [batch, shape[...]]

  """
  multiples = tf.concat(
      [tf.constant([batch]),
       tf.ones([len(shape)], dtype=tf.int32)], axis=-1)

  return tf.tile(
      tf.expand_dims(tf.scatter_nd(indices, values, shape), axis=0), multiples)


def generate_extrinsic(yaw, pitch, roll, translation, batch=1):
  """Generates extrinsic.

  Args:
    yaw: scalar tensor
    pitch: scalar tensor
    roll: scalar tensor
    translation: [3] tensor
    batch: integer

  Returns:
    [batch, 4, 4] tensor
  """
  rotation_matrix = transform_utils.get_rotation_matrix(roll, pitch, yaw)
  return tf.tile(
      tf.expand_dims(
          transform_utils.get_transform(
              rotation_matrix, tf.constant(translation, dtype=tf.float32)),
          axis=0), [batch, 1, 1])

File: waymo_open_dataset/metrics/test_utils.cc
/* Copyright 2019 The Waymo Open Dataset Authors. All Rights Reserved.

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

    http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
==============================================================================*/

#include "waymo_open_dataset/metrics/test_utils.h"

#include "waymo_open_dataset/label.pb.h"
#include "waymo_open_dataset/protos/breakdown.pb.h"
#include "waymo_open_dataset/protos/metrics.pb.h"

namespace waymo {
namespace open_dataset {

Label::Box BuildAA2dBox(double center_x, double center_y, double length,
                        double width) {
  Label::Box box;
  box.set_center_x(center_x);
  box.set_center_y(center_y);
  box.set_length(length);
  box.set_width(width);
  return box;
}

Label::Box BuildBox2d(double center_x, double center_y, double length,
                      double width, double heading) {
  Label::Box box = BuildAA2dBox(center_x, center_y, length, width);
  box.set_heading(heading);
  return box;
}

Label::Box BuildBox3d(double center_x, double center_y, double center_z,
                      double length, double width, double height,
                      double heading) {
  Label::Box box = BuildBox2d(center_x, center_y, length, width, heading);
  box.set_center_z(center_z);
  box.set_height(height);
  return box;
}

Polygon2dProto BuildNLZ(const std::vector<double>& x,
                        const std::vector<double>& y) {
  Polygon2dProto nlz;
  for (int i = 0, sz = x.size(); i < sz; ++i) {
    nlz.add_x(x[i]);
    nlz.add_y(y[i]);
  }
  return nlz;
}

Config BuildDefaultConfig() {
  Config config;
  config.mutable_score_cutoffs()->Reserve(10);
  for (int i = 0; i < 10; ++i) {
    config.add_score_cutoffs(0.1 * i);
  }
  config.add_breakdown_generator_ids(Breakdown::ONE_SHARD);
  config.add_difficulties();

  config.set_matcher_type(MatcherProto::TYPE_HUNGARIAN_TEST_ONLY);
  for (int i = 0; i <= Label::Type_MAX; ++i) {
    config.add_iou_thresholds(0.5);
  }
  config.set_box_type(Label::Box::TYPE_3D);
  return config;
}

}  // namespace open_dataset
}  // namespace waymo
